harness/determined/pytorch/_pytorch_context.py

```diff
@@ -17,11 +32,23 @@ class PyTorchTrialContext(det.TrialContext):
     def __init__(self, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
 
+        self._init_device()
+        # Track whether a warning logging category has already been issued to the user.
+        self.warning_logged = {_WarningLogs.FAILED_MOVING_TO_DEVICE: False}
+
         # The following three attributes are initialized during the lifetime of
         # a PyTorchTrialContext.
         self.model = None  # type: Optional[nn.Module]
-        self.optimizer = None
-        self.lr_scheduler = None  # type: Optional[pytorch.LRScheduler]
+        self.optimizers = []  # type: List[torch.optim.Optimizer] #  type: ignore
+        self.lr_schedulers = []  # type: List[pytorch.LRScheduler]
+        self.callbacks = {}  # type:  Dict[str, pytorch.PyTorchCallback]
+
+        self.scaler_mapping = {}  # type: Dict[torch.Tensor, int]
+        self.loss_num = 0
```

Comment:
blocking: rename to something a bit clearer, like `next_loss_id` or `next_apex_loss_id`